src/backend/renderer/gles2/mod.rs

```diff
@@ -1283,15 +1283,24 @@ impl Frame for Gles2Frame {
             (texture_mat * Vector3::new(src.loc.x as f32, (src.loc.y + src.size.h) as f32, 0.0)).truncate(), // bottom-left
         ];
 
+        let damage_viewport = Rectangle::from_extemities((0f64, 0f64), dest.size.to_point());
+
         let damage = damage
             .iter()
-            .map(|rect| {
-                [
-                    rect.loc.x as f32 / dest.size.w as f32,
-                    rect.loc.y as f32 / dest.size.h as f32,
-                    rect.size.w as f32 / dest.size.w as f32,
-                    rect.size.h as f32 / dest.size.h as f32,
-                ]
+            .flat_map(|rect| {
+                let damage = rect
+                    .to_f64()
+                    .intersection(damage_viewport)
+                    .map(|rect_constrained| {
+                        [
+                            (rect_constrained.loc.x / dest.size.w) as f32,
+                            (rect_constrained.loc.y / dest.size.h) as f32,
+                            (rect_constrained.size.w / dest.size.w) as f32,
+                            (rect_constrained.size.h / dest.size.h) as f32,
+                        ]
+                    });
+                debug_assert!(damage.is_some());
```

Comment:
Why this debug assert? Do we make it a fatal error to provide damage that is outside the viewport?

Reply:
The damage is what is used for rendering in `render_texture` defining the instances for instanced rendering. If the damage is outside of the `damage_viewport` in the worst case nothing will be drawn. Imo this is an error in the calling code but may be hard to find which this `debug_assert!` aims to make easier.

Reply:
what's your opinion on that @Drakulix ?

Reply:
Our code should never do this and it would be nice to catch cases early.

I would say we leave it in, until someone opens an issue with a legit use-case for providing non-overlapping damage values. Because I can't think of any.

Reply:
Alright